### PR TITLE
Hyper upgrade staging

### PR DESCRIPTION
**Pull request by garypen**
The upgrade is ready for review.

Please read the following documents **before** starting your review:
 
 - HYPER_1.0_REVIEW_NOTES.md
 - HYPER_1.0_UPDATE.md

There are also these provisions:
 - Not all tests are passing or converted yet
 - We know there's a graceful shutdown problem we need to address

Thanks for the reviews.

Ref: #3775 

[ROUTER-7]: https://apollographql.atlassian.net/browse/ROUTER-7?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by apollo-bot2** [comment]
## Detected SAST Vulnerabilities
## :red_circle: Potential Security Issues Found

If you are seeing this message, it means that [the security scanning tool](https://semgrep.dev/) that Apollo uses to test our source has identified a potential security issue in code introduced or updated via your branch. Information about what was detected and steps to move forward are below.

If the scanner detected a legitimate issue, please take action to correct it prior to merging this PR. The action required will vary based on the detection. If the detection is a false positive, please follow the steps below to resolve the issue.

###  Issues Detected 
| Message | Caused CI Failure? | Link | Path | Signature |
|---------|--------------------|------|------|-----------|
| Insecure WebSocket Detected. WebSocket Secure (wss) should be used for all WebSocket connections. | Yes | <a href="https://github.com/apollographql/router/blob/c6fa0aa7c162893712db29febb018475c594f261/apollo-router/src/protocols/websocket.rs?plain=1#L967" target="_blank">Link</a> | apollo-  router/src/protocols/websocket.rs | rules.providers.semgrep.security.ja  vascript.lang.security.detect-  insecure-websocket |
| Insecure WebSocket Detected. WebSocket Secure (wss) should be used for all WebSocket connections. | Yes | <a href="https://github.com/apollographql/router/blob/c6fa0aa7c162893712db29febb018475c594f261/apollo-router/src/protocols/websocket.rs?plain=1#L1034" target="_blank">Link</a> | apollo-  router/src/protocols/websocket.rs | rules.providers.semgrep.security.ja  vascript.lang.security.detect-  insecure-websocket |
<hr />
<details>
    <summary><h3>False Positive Resolution Process</h3></summary>

The easiest way to resolve a false positive is to add a comment containing `nosemgrep` to the code that triggered the detection. You can add the comment directly to the line triggering the detection or on the line immediately above it. For example:

```python
#!/usr/bin/env python3

def my_function():
    my_code_generating_detection() # nosemgrep
```

<details>
    <summary><h4>What if I can't add a line comment?</h4></summary>

If you can't add a comment because the detection is generated by a file that does not support comments (like JSON), you can use Semgrep's `.semgrepignore` file. For more information about how to create an exclusion via `.semgrepignore`, refer to the [Semgrep documentation](https://semgrep.dev/docs/ignoring-files-folders-code/#defining-ignored-files-and-folders-in-semgrepignore).
</details>

</details>
<hr />
<details> 
    <summary><h3>Getting Help</h3></summary>


The Apollo Security team is available to assist in resolving this issue. Please tag us on this PR using  `@apollographql/security` if you need assistance! 
</details>
<hr />
<details>
    <summary><h3>How do I know I fixed this correctly?</h3></summary>
Once you have resolved the issue completely, this message will disappear! If you're still seeing this message, there is more to do prior to merging.
</details>

File: apollo-router/src/services/router/body.rs
Comment (lrlna) [inline apollo-router/src/services/router/body.rs:37]: `from_data_stream` to match its opposite function, `into_data_stream`? 
```suggestion
pub(crate) fn from_data_stream<S, E>(data_stream: S) -> RouterBody
```

File: apollo-router/src/plugins/traffic_shaping/mod.rs
Comment (lrlna) [inline apollo-router/src/plugins/traffic_shaping/mod.rs:450]: ... the variants used to be named ... `A` and `B`? :scream: that's ... very descriptive :scream:

File: apollo-router/src/plugins/traffic_shaping/retry.rs
Comment (lrlna) [inline apollo-router/src/plugins/traffic_shaping/retry.rs:89]: why did we remove the test below?

File: apollo-router/src/services/router/body.rs
Comment (lrlna) [inline apollo-router/src/services/router/body.rs:17]: should this be `into_bytes`?
```suggestion
pub(crate) async fn into_bytes<B: HttpBody>(body: B) -> Result<Bytes, B::Error> {
    Ok(body.collect().await?.to_bytes())
}
```

File: apollo-router/src/services/router/body.rs
Comment (lrlna) [inline apollo-router/src/services/router/body.rs:50]: should this be `into_string`?
```suggestion
pub(crate) async fn into_string<B>(input: B) -> Result<String, AxumError>
```

File: apollo-router/src/services/router.rs
Comment (lrlna) [inline apollo-router/src/services/router.rs:73]: can we use `#[cfg(test)]` around this? And should this be `pub` or `pub(crate)`?

File: apollo-router/src/services/subgraph_service.rs
Comment (lrlna) [inline apollo-router/src/services/subgraph_service.rs:1961]: this is technically not part of this changeset, but should this be `handler`? `listener` + `handler` seem to be more aligned than `listener` + `handle`

File: apollo-router/tests/integration/file_upload.rs
Comment (lrlna) [inline apollo-router/tests/integration/file_upload.rs:890]: hmmm this is missing the more important part of the error. Where did it go?

**Review by lrlna — COMMENTED** [review]
Whoa, this was a lot! Good job everyone on getting this through! I went through the whole thing, and here are some things I observed.

File: apollo-router/benches/deeply_nested.rs
Comment (garypen) [reply apollo-router/benches/deeply_nested.rs:142]: The "legacy" client is the client from hyper 0.14. This is what we use in router 1.0. I think @bnjjj started to figure out what was required to use the new client builder and it wasn't straightforward.
For now, the least disruptive thing to do was to use, effectively, the same Client. I think that changing this decision in a follow up PR should be fine.

File: apollo-router/src/services/router/service.rs
Comment (garypen) [reply apollo-router/src/services/router/service.rs:342]: I think that is an artefact of "different people doing the same thing at the same time in different ways". I did try to standardise things as much as possible, but missed this one.

File: apollo-router/src/services/router.rs
Comment (garypen) [reply apollo-router/src/services/router.rs:73]: It's being used in integration tests. I'll update the comment to make that clearer.

File: apollo-router/tests/integration/file_upload.rs
Comment (garypen) [reply apollo-router/tests/integration/file_upload.rs:890]: This is one of the bad side-effects of the changes in Hyper regarding Body implementation. We still hope to resolve this either before the PR is merged or shortly after.

File: apollo-router/src/services/router/body.rs
Comment (garypen) [reply apollo-router/src/services/router/body.rs:35]: Yup. Naming does need a bit of polishing. I like your suggestion, and in such a scheme, this would be `from_bytes` I think.

File: apollo-router/src/services/subgraph_service.rs
Comment (goto-bus-stop) [reply apollo-router/src/services/subgraph_service.rs:1961]: i think there is no upside or downside to either `handle` or `handler`.
the listener is an object that is already listening. only a noun works.
`handle` is a function that will be called to handle a request, so a verb works. it's also a thing that handles a request so a noun works too. `handler` might be more aligned but there is no particular reason for these to be aligned.

File: apollo-router/src/plugins/traffic_shaping/mod.rs
Comment (garypen) [reply apollo-router/src/plugins/traffic_shaping/mod.rs:450]: That's an underlying crate name change.

File: apollo-router/src/services/router/body.rs
Comment (garypen) [reply apollo-router/src/services/router/body.rs:50]: done: 555606addd96cd8bb44af2fe00a44b0a08d2d671

File: apollo-router/src/services/router/body.rs
Comment (garypen) [reply apollo-router/src/services/router/body.rs]: done: 555606addd96cd8bb44af2fe00a44b0a08d2d671

File: apollo-router/src/services/router/body.rs
Comment (garypen) [reply apollo-router/src/services/router/body.rs:37]: I'm not going to make this change since it actually is a result stream that we are processing.

File: apollo-router/src/plugins/traffic_shaping/retry.rs
Comment (garypen) [reply apollo-router/src/plugins/traffic_shaping/retry.rs:89]: That feature is to be removed (soon) and so the test wasn't ported. I've ported it and added it back so the test can be removed when the feature is removed.

File: apollo-router/src/services/router.rs
Comment (garypen) [reply apollo-router/src/services/router.rs:73]: done: 27219bb5a8129882705f17888f745a89be5d67dd

File: apollo-router/src/services/router/service.rs
Comment (garypen) [reply apollo-router/src/services/router/service.rs]: done: e8c01704f38d5fcf286a66b0c82f8fd90bdc538c

File: apollo-router/src/services/http/tests.rs
Comment (garypen) [reply apollo-router/src/services/http/tests.rs]: We are inconsistent in quite a few places actually:
```
apollo-router/src/batching.rs
apollo-router/src/plugins/authentication/tests.rs
apollo-router/src/services/router/tests.rs
apollo-router/src/services/router/service.rs
apollo-router/src/services/subgraph_service.rs
apollo-router/src/services/external.rs
apollo-router/src/services/router.rs
apollo-router/src/services/http/tests.rs
apollo-router/src/plugins/telemetry/mod.rs
apollo-router/src/plugins/limits/mod.rs
apollo-router/src/plugins/coprocessor/supergraph.rs
apollo-router/src/plugins/coprocessor/mod.rs
apollo-router/src/plugins/coprocessor/execution.rs
apollo-router/src/plugins/coprocessor/test.rs
apollo-router/src/plugins/cache/invalidation_endpoint.rs
```
I'll apply some consistency...

Comment (garypen) [reply apollo-router/src/services/http/tests.rs]: Ok: I've consistently applied these rules:
```
import crate::services::router;
import crate::services::router::body::RouterBody;
```

Use functions as:
```
router::body::into_bytes
router::body::from_bytes
```
etc...

done: ae71ed8e2e8a6aea92d3f6ee4e1edbeb09c683c3

File: HYPER_1.0_REVIEW_NOTES.md
Comment (garypen) [reply HYPER_1.0_REVIEW_NOTES.md:135]: I was going to leave them in the repo for now. If we ever do remove them, then putting the contents into confluence would be a good idea.

File: apollo-router/tests/integration/file_upload.rs
Comment (IvanGoncharov) [inline apollo-router/tests/integration/file_upload.rs:691]: nice, -1 wrapper 🎉

**Review by goto-bus-stop — APPROVED** [review]
Un-biased approval

File: apollo-router/src/axum_factory/axum_http_server_factory.rs
Comment (garypen) [reply apollo-router/src/axum_factory/axum_http_server_factory.rs:447]: I'm going to merge. I've raised a follow-up issue for this.

**Pull request merged**
